### PR TITLE
Rawls submits Dockstore WDLs to Cromwell by URL [WA-68]

### PR DESCRIPTION
**Pull request by aednichols**
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by aednichols** [comment]
Good news, I ran `POST /api/template` on my FiaB with payload
```
{
  "methodUri": "dockstore://github.com%2Fdockstore-testing%2Fmd5sum-checker%2Fwdl_wdl_checker/1.0.1"
}
```
and got back
```
{
  "deleted": false,
  "inputs": {
    "checkerWorkflow.expectedMd5sum": "",
    "checkerWorkflow.inputFile": ""
  },
  "methodConfigVersion": 1,
  "methodRepoMethod": {
    "methodUri": "dockstore://github.com%2Fdockstore-testing%2Fmd5sum-checker%2Fwdl_wdl_checker/1.0.1",
    "sourceRepo": "dockstore",
    "methodPath": "github.com/dockstore-testing/md5sum-checker/wdl_wdl_checker",
    "methodVersion": "1.0.1"
  },
  "name": "name",
  "namespace": "namespace",
  "outputs": {
    "checkerWorkflow.md5call.value": ""
  },
  "prerequisites": {},
  "rootEntityType": "rootEntityType"
}
```

That payload is the [example relative-import workflow provided by Dockstore](https://dockstore.org/workflows/github.com/dockstore-testing/md5sum-checker/wdl_wdl_checker:1.0.1?tab=info).

**Comment by aednichols** [comment]
jenkins retest

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala [hunk 1]
```diff
@@ -22,8 +22,13 @@ class CromwellSwaggerClient(cromwellBasePath: String) extends LazyLogging {
...
-  def describe(userInfo: UserInfo, wdl: String): Try[WorkflowDescription] = {
-    Retry.retry(5.seconds, 30.seconds) { Try { getCromwellWomtoolApi(userInfo.accessToken.token).describe("v1", wdl, null, null, null, null) } }
-  }
+  def describe(userInfo: UserInfo, wdl: WDL): Try[WorkflowDescription] =
+    wdl.url match {
+      case Some(url) =>
+        Retry.retry(5.seconds, 30.seconds) { Try { getCromwellWomtoolApi(userInfo.accessToken.token).describe("v1", null, url, null, null, null) } }
+      case None =>
+        Retry.retry(5.seconds, 30.seconds) { Try { getCromwellWomtoolApi(userInfo.accessToken.token).describe("v1", wdl.source, null, null, null, null) } }
+    }
```
Comment (kshakir) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala]: Reading this, it seems like a `WDL` is an "or" relation versus an "and" relation...

WdlSource OR WdlUrl, not WdlSource AND WdlUrl?

Thus would a sealed trait make more sense than a case class / Product?

**Review by kshakir — COMMENTED** [review]
A question about the WDL ADT, but otherwise the general code looks good.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala (hunk 1 above)
Comment (aednichols) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala]: See comment in `WDL.scala`, they're not designed to be exclusive, rather to coexist and be used in complement with each other.
```
 // The kernel of Cromwell's `cromwell.core.WorkflowSourceFilesCollection` - a workflow and its location, to enable relative imports
```

**Review by kshakir — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala (hunk 1 above)
Comment (aednichols) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala]: Just kidding, your idea proved extremely good

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -2328,11 +2328,11 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
 
 
   private def withSubmissionParameters(workspaceName: WorkspaceName, submissionRequest: SubmissionRequest)
-    (op: (DataAccess, SlickWorkspaceContext, String, SubmissionValidationHeader, Seq[SubmissionValidationEntityInputs], Seq[SubmissionValidationEntityInputs], Option[WorkflowFailureMode]) => ReadWriteAction[PerRequestMessage]): Future[PerRequestMessage] = {
+    (op: (DataAccess, SlickWorkspaceContext, SubmissionValidationHeader, Seq[SubmissionValidationEntityInputs], Seq[SubmissionValidationEntityInputs], Option[WorkflowFailureMode]) => ReadWriteAction[PerRequestMessage]): Future[PerRequestMessage] = {
     dataSource.inTransaction { dataAccess =>
       withWorkspaceContextAndPermissions(workspaceName, SamWorkspaceActions.write, dataAccess) { workspaceContext =>
         withMethodConfig(workspaceContext, submissionRequest.methodConfigurationNamespace, submissionRequest.methodConfigurationName, dataAccess) { methodConfig =>
-          withMethodInputs(methodConfig, userInfo) { (wdl, gatherInputsResult) =>
+          withMethodInputs(methodConfig, userInfo) { gatherInputsResult =>
```
Comment (aednichols) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:2335]: This change is a bit of a stowaway, it's not necessary for the feature to work. Simply, the `wdl` part of this tuple is not used anywhere `withMethodInputs` is called, so I removed it from the return type.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockCromwellSwaggerClient.scala [hunk 3]
```diff
@@ -11,9 +11,17 @@ import scala.util.Try
...
 
-  val workflowDescriptions: mutable.Map[String, WorkflowDescription] =  new TrieMap()
-
-  override def describe(userInfo: UserInfo, wdl: String): Try[WorkflowDescription] = {
+  val workflowDescriptions: mutable.Map[WDL, WorkflowDescription] =  new TrieMap()
+
+  override def describe(userInfo: UserInfo, wdl: WDL): Try[WorkflowDescription] = {
+    if (!workflowDescriptions.contains(wdl)) {
+      throw new Exception(
+        "Danger! Possible misconfigured test: the MockCromwellSwaggerClient received a request for a WDL it doesn't have and will return an error. " +
```
Comment (aednichols) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockCromwellSwaggerClient.scala:19]: `total_hours_wasted_here = 42`

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolverSpec.scala [hunk 4]
```diff
@@ -18,121 +18,127 @@ class MethodConfigResolverSpec extends WordSpecLike with Matchers with TestDrive
...
+        |    File f2 = "a"
+        |  }
+        |
+        |  runtime {
+        |    docker: "docker image"
+        |  }
+        |}
+      """.stripMargin)
+
+  val badWdl = WdlSource("This is not a valid workflow [MethodConfigResolverSpec]")
```
Comment (aednichols) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolverSpec.scala:141]: There are so many WDLs flying around in these test suites that I had to label where they came from to make any sense of it

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala [hunk 5]
```diff
@@ -21,9 +21,16 @@ class CromwellSwaggerClient(cromwellBasePath: String) extends LazyLogging {
...
-    Retry.retry(5.seconds, 30.seconds) { Try { getCromwellWomtoolApi(userInfo.accessToken.token).describe("v1", wdl, null, null, null, null) } }
-  }
+  def describe(userInfo: UserInfo, wdl: WDL): Try[WorkflowDescription] =
+    Retry.retry(5.seconds, 30.seconds) {
+      Try {
+        wdl match {
+          case wdl: WdlUrl =>
+            getCromwellWomtoolApi(userInfo.accessToken.token).describe("v1", null, wdl.url, null, null, null)
+          case wdl: WdlSource =>
+            getCromwellWomtoolApi(userInfo.accessToken.token).describe("v1", wdl.source, null, null, null, null)
```
Comment (aednichols) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala:31]: For those in a hurry, this is the part that really does the thing

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala [hunk 6]
```diff
@@ -40,12 +40,18 @@ class HttpExecutionServiceDAO(executionServiceURL: String, override val workbenc
...
+  override def submitWorkflows(wdl: WDL, inputs: Seq[String], options: Option[String], labels: Option[Map[String, String]], workflowCollection: Option[String], userInfo: UserInfo): Future[Seq[Either[ExecutionServiceStatus, ExecutionServiceFailure]]] = {
     val url = executionServiceURL+"/api/workflows/v1/batch"
     val labelsBodyPart = labels.map(labelsMap => Multipart.FormData.BodyPart("labels",labelsMap.toJson.compactPrint))
     val wfc = workflowCollection.map(Multipart.FormData.BodyPart("collectionName", _))
 
-    val bodyParts = Seq(Multipart.FormData.BodyPart("workflowSource", wdl),
+    val wdlSourceOrUrl = wdl match {
+      case wdl: WdlUrl => Multipart.FormData.BodyPart("workflowUrl", wdl.url)
+      case wdl: WdlSource => Multipart.FormData.BodyPart("workflowSource", wdl.source)
+    }
```
Comment (aednichols) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala:51]: Equivalent doing of the thing in the second of two places

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpMethodRepoDAO.scala [hunk 7]
```diff
@@ -53,18 +53,20 @@ class HttpMethodRepoDAO(agoraConfig: MethodRepoConfig[Agora.type], dockstoreConf
...
+          for {
+            entity: AgoraEntity <- maybeEntity
+            payload <- entity.payload
+          } yield WdlSource(payload)
+        }
       case dockstoreMethod: DockstoreMethod =>
-        getDockstoreMethod(dockstoreMethod) flatMap { response: Option[GA4GHTool] =>
-          response match {
-            case Some(tool) =>
-              // TODO: re-using AgoraEntity feels sketchy to me. It seems to work without any changes, but should we create a DockstoreEntity?
```
Comment (aednichols) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpMethodRepoDAO.scala:64]: TFW you kill a TODO you left for yourself in 2018

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockCromwellSwaggerClient.scala (hunk 3 above)
Comment (rsasch) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockCromwellSwaggerClient.scala:19]: worth it 🥇

**Review by rsasch — APPROVED** [review]
LGTM

**Review by kshakir — APPROVED** [review]

**Review by salonishah11 — APPROVED** [review]

**Pull request merged**